### ISSUE TITLE
append mode in status view

### ISSUE THREAD

**Issue by Avinash-Bhat** [jonas/tig#211]
While editing a commit in tig, it is not apparent what the end commit will be if we are to append the edits. So it's better to have an append mode in the status view.

Once the append mode is activated, tig should show the diff with `HEAD`'s parent. Activating the commit keybinding will activate <del>`commit -a`</del> `commit --amend` instead of `commit`.

**Comment by CodeMan99** [comment]
Why `--amend` at all? Just `git reset HEAD^` before you start working. I mean, what disadvantage would there be?

**Comment by Avinash-Bhat** [comment]
@cody,

Resetting will make you lose the commit message for one.
In amend mode you can do selective reset.
To make this more clear, try using the amend mode in git-cola.

**Comment by CodeMan99** [comment]
Commit message are intended to be be short and easy to remember. More often than not your commit message changes anyway. If your commits are really large enough to justify a selective reset as reason to use `--amend` I propose that tig doesn't fit your workflow in the first place.

**Comment by CodeMan99** [comment]
You could also bind `git diff --staged` for this purpose.

**Comment by Avinash-Bhat** [comment]
Yes they are also intended to be explaining the changes. The first line of
the commit should be short, however there can be subsequent  lines
explaining why things are done in the commit. This kind of commit messages
are seen in Linux kernel tree and others.

Yes there might be workarounds for this. But this may not be apparent for
everyone.

I think this is a really cool feature to have with tig.

—
Reply to this email directly or view it on
<https://github.com/jonas/tig/issues/211#issuecomment-70036606>GitHub
<https://github.com/jonas/tig/issues/211#issuecomment-70036606>.<img src="
https
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
://ci5.
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
googleusercontent.com
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
-
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
Kd0xyWupLkJfX__aA1tHqTWvgu
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
-5p8aplM-s44m-
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
qbnXDqbDg6D5xcBbgywrOJMig
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
=s0-d-e1-ft#
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
https
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
://
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
github.com
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
/notifications/beacon/
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
ABZp7r_W
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
-
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif
<https://ci5.googleusercontent.com/proxy/4SZWL5rcwjJMsCHIgqzDaVLQYnxH6hGyepM1crgL_pWGzzx_Jlfjbw1c-n0TKEqQ1iabmvBc9JWvKYm7jNJDTTf_1VX8-Kd0xyWupLkJfX__aA1tHqTWvgu-5p8aplM-s44m-qbnXDqbDg6D5xcBbgywrOJMig=s0-d-e1-ft#https://github.com/notifications/beacon/ABZp7r_W-i5iSQACPS7821YwcAg3Zimjks5nhzHZgaJpZM4BFYnE.gif>
">

**Comment by Avinash-Bhat** [comment]
@cody, consider the following scenario:
You made a commit that should've been two commits. How to go about
splitting this?

With tig amend mode, you could do this without resetting the tree. (This is
possible in cola)

**Comment by CodeMan99** [comment]
I do `git rebase -i HEAD~2` on the command line. While tig is great, it's workflow is simple. If you want a direct feature of git, use git.

For the original problem I still suggest binding something to `git diff --staged %(commit)`. This will cover all use cases for `--staged`:
 1. See only staged changes - select the HEAD commit in main view.
 2. See all the changes if you were to `commit --amend` - select the HEAD^ commit in the main view.
 3. See if it's reasonable to squash the index into two or more previous commits - select the HEAD~(N) commit you don't want to include.

**Comment by CodeMan99** [comment]
@Avinash-Bhat I've stumbled across another option. After doing `git reset --soft` you will have an `ORIG_HEAD` pointer that keeps the commits you just reset. You can use that like `git commit -c ORIG_HEAD` to keep the same commit information.

Seeing as how @jonas wrote tig mostly for his own use usually writes features he will use himself (correct if I'm wrong here @jonas). This seems like an edge case with more than one work around, so creating a specialized mode for appending to a commit seems isn't very likely.

**Comment by studgeek** [comment]
I think amend is a common workflow for many (e.g. Android AOSP development using Gerrit). I believe this is partially evidenced by the fact that git team added the --amend feature in the first place.

**Comment by nidico** [comment]
I agree this is a missing feature. I just switched from `git-cola` to `tig` for commiting and this is the only feature I'm missing so far.
